### PR TITLE
Don't use queries when printing the query stack

### PR DESCRIPTION
**Pull request by Zoxc**
This adds a parameter to `try_collect_active_jobs` indicating if it can use queries to describe query frames. This is set to false when when printing the query stack during an ICE to work around the infinite recursion seen in https://github.com/rust-lang/rust/issues/112522.

### CODE REVIEW

**Comment by compiler-errors** [comment]
This seems like a really heavy hammer to solve this issue.

Is it possible to just use `with_no_queries`/`NO_QUERIES` more judiciously within the pretty-printing code? I'd rather we print *something* for the query stack -- knowing the query frame descr is often very useful...

File: compiler/rustc_query_impl/src/plumbing.rs [hunk 1]
```diff
@@ -300,7 +300,13 @@ pub(crate) fn create_query_frame<
     key: K,
     kind: DepKind,
     name: &'static str,
+    can_call_queries: bool,
 ) -> QueryStackFrame<DepKind> {
+    if !can_call_queries {
+        // Return a minimal query frame if we can't call queries
+        return QueryStackFrame::new(String::new(), None, None, None, kind, None, || Hash64::ZERO);
+    }
```
Comment (bjorn3) [inline compiler/rustc_query_impl/src/plumbing.rs:308]: Can this be moved below the `description` formatting? It seems like that shouldn't use any queries.

**Comment by Zoxc** [comment]
It would be nice to at least print query keys, but I think we need either a separate non-query printing path or a better way to statically prevent queries to run for the existing path.

File: compiler/rustc_query_impl/src/plumbing.rs (hunk 1 above)
Comment (Zoxc) [reply compiler/rustc_query_impl/src/plumbing.rs:308]: That can and does use queries.

**Comment by eholk** [comment]
@compiler-errors - is the root of your concern that we're losing a lot of really helpful query information in the common case to avoid recursion in pathological cases?

I wonder if having a `recursion_depth` parameter we thread around would be a better compromise?

**Comment by compiler-errors** [comment]
> is the root of your concern that we're losing a lot of really helpful query information in the common case to avoid recursion in pathological cases?

Pretty much, yeah. The query stack being able to print keys on ICEs, for example, is quite helpful in my experience debugging the compiler.

**Comment by eholk** [comment]
@Zoxc - Do you think a recursion limit would work?

@rustbot author

**Comment by Zoxc** [comment]
A recursion limit doesn't help for the parallel compiler which has more interesting failure scenarios. A more reliable way to print def paths is probably the way to go to restore debugging information.

**Comment by cjgillot** [comment]
This PR is very complex, and makes the query description weird by changing the key type behind our back.
Alternative version: could we remove access to the tcx from TLS when we are in the plumbing code?

Today, we manipulate TLS twice when setting up a query:
- once in `start_query` to setup the recursion, token and side-effect trackers,
- once in `with_task` to setup dependency tracking.

Can the outer one remove access to the tcx, and have the inner one put it back?

**Comment by JohnCSimon** [comment]
@Zoxc 

ping from triage - can you post your status on this PR? This PR has not received an update in a few months. Thank you!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by Zoxc** [comment]
It doesn't seem like either approaches taken in this PR are very popular.

We may instead make the existing printing code respect `with_no_queries!`.

**Pull request closed without merging**